README.md
# clj-base-crm

[![Clojars](https://img.shields.io/clojars/v/clj-base-crm.svg)](http://clojars.org/clj-base-crm)

clj-base-crm is a simple client library for accessing the Base CRM API.

Please read through the [Base CRM API Endpoint Reference](https://developers.getbase.com/docs/rest/articles/introduction).

## Installation

Add the following to your project.clj dependencies:

```clj
[clj-base-crm "0.2.0"]
```

## Documentation

[API Docs](http://ryanbertrand.github.io/clj-base-crm/clj-base-crm.core.html)


## Usage

Require the `clj-base-crm.core` in your namespace:

```clj
(require '[clj-base-crm.core :as base])
```

Setup your Base Personal Access Tokens (PAT) [Don't have one? Read this](https://developers.getbase.com/docs/rest/articles/first_call).

```clj
(base/set-access-token! "YOUR_BASE_PAT_ACCESS_TOKEN")
```

clj-base-crm only supports the following resources:

```clj
::base/leads
::base/lead-sources
::base/contacts
::base/notes
```

To create a new lead, use the `create` function.

```clj
(base/create ::base/leads
 {:first_name "John" :last_name "Doe" :email "john.doe@fakemail.com"})
```

To retrieve a new lead by it's ID, use the `retrieve` function.

```clj
(base/retrieve ::base/leads {:id 1234})
```

To update a lead by it's ID, use the `update` function.

```clj
(base/update ::base/leads {:id 1234} {:email "newemail@fakemail.com"})
```

To delete a lead by it's ID, use the `delete` function.

```clj
(base/delete ::base/leads  {:id 1234})
```

Sometimes you need to create a lead but you are not sure if this lead is already in the system. Base provides an upsert action which will create a new lead or update an existing lead based on the filters you pass in.
To upsert a lead by it's ID, use the `upsert` function.

This will upsert based on the email filter `john.doe@fakemail.com`.
```clj
(base/upsert ::base/leads
             {:email "john.doe@fakemail.com"}
             {:first_name "John" :last_name "Doe" :phone "916-456-7890"})
```

This will upsert based on the custom field filter `external_id` `123`.
```clj
(base/upsert ::base/leads
             {:custom_fields {:external_id 123}}
             {:first_name "John" :last_name "Doe" :phone "916-456-7890"})
```

## Return values

clj-base-crm returns the data received from the Base CRM API unaltered but the response will be converted from json to a Clojure map.

##Error handling

clj-base-crm does not throw exceptions on exceptional status codes from the Base API.  It returns the error body.

##Todo
clj-base-crm only exposes a few resources.  I plan to add more as I need them.  If you would like to expose more (just a simple hash-map), feel free to submit a PR!

- Support more Base resources
- Tests

## License

Copyright (c) 2016 Ryan Bertrand

Distributed under the Eclipse Public License either version 1.0 or (at your option) any later version.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-04-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-04-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/clj-base-crm/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj-base-crm/compare/0.1.0...0.1.1


test/clj_base_crm/core_test.clj
(ns clj-base-crm.core-test
  (:require [clojure.test :refer :all]
            [clj-base-crm.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj_base_crm/leads.clj
(ns clj-base-crm.leads
  (:refer-clojure :exclude [update])
  (:require [clj-base-crm.core :as base]))

;API Docs: https://developers.getbase.com/docs/rest/reference/leads

(def endpoint "/v2/leads")

(defn retrieve-all
  "Retrieve all leads according to the parameters"
  ([]
   (base/retrieve-all nil))
  ([{:keys [page per_page sort_by ids creator_id owner_id source_id] :as parameters}]
   (base/retrieve-all endpoint parameters)))

(defn retrieve
  "Retrieve a single lead"
  [id]
  (base/retrieve (format "%s/%s" endpoint id)))

(defn create
  "Creates a new lead"
  [data]
  (base/create endpoint data))

(defn update
  "Updates lead information"
  [id data]
  (base/update (format "%s/%s" endpoint id) data))

(defn delete
  "Delete an existing lead"
  [id]
  (base/delete (format "%s/%s" endpoint id)))

(defn upsert
  "Create a new lead or update an existing, based on a value of a filter
   or a set of filters. At least one filter is required."
  [filters data]
  (base/upsert (format "%s/upsert" endpoint) filters data))

src/clj_base_crm/contacts.clj
(ns clj-base-crm.contacts
  (:refer-clojure :exclude [update])
  (:require [clj-base-crm.core :as base]))

;API Docs: https://developers.getbase.com/docs/rest/reference/contacts

(def endpoint "/v2/contacts")

(defn retrieve-all
  "Retrieve all contacts according to the parameters"
  ([]
   (base/retrieve-all nil))
  ([{:keys [page per_page sort_by ids creator_id owner_id source_id] :as parameters}]
   (base/retrieve-all endpoint parameters)))

(defn retrieve
  "Retrieve a single contact"
  [id]
  (base/retrieve (format "%s/%s" endpoint id)))

(defn create
  "Creates a new contact"
  [data]
  (base/create endpoint data))

(defn update
  "Updates contact information"
  [id data]
  (base/update (format "%s/%s" endpoint id) data))

(defn delete
  "Delete an existing contact"
  [id]
  (base/delete (format "%s/%s" endpoint id)))

(defn upsert
  "Create a new contact or update an existing, based on a value of a filter
   or a set of filters. At least one filter is required."
  [filters data]
  (base/upsert (format "%s/upsert" endpoint) filters data))

src/clj_base_crm/core.clj
(ns clj-base-crm.core
  (:refer-clojure :exclude [update])
  (:require [clj-http.client :as http]
            [clojure.string :as s]))

;API Docs: https://developers.getbase.com/docs/rest/articles/first_call

(def ^:no-doc base-url "https://api.getbase.com")

;Base CRM API Token
(def ^:no-doc ^:dynamic *access-token* "YOUR_BASE_PAT_ACCESS_TOKEN")

(defn set-access-token!
  "Set the Access Token to be used by default on all requests."
  [token]
  (def ^:dynamic *access-token* token))

(defmacro with-token
  "Allows you to make API calls using this access token instead of the global token.
  Useful for many users scenarios."
  [token & body]
  `(binding [*access-token* ~token]
     (do ~@body)))

(defn ^:no-doc base-api-call
  "Sends a HTTP request to the Base CRM REST API.
  Returns the called URL and the response status in the metadata."
  [method partial-url & opts]
  (let [opts (into {} opts)
        url (str base-url partial-url)
        _ (println url)
        req (merge {:url                   url
                    :method                method
                    :oauth-token           *access-token*
                    :headers               {:accept "application/json"}
                    :coerce                :always
                    :as                    :json
                    :throw-entire-message? true
                    :content-type          :json} opts)
        {:keys [status body] :as response} (try
                                             (http/request req)
                                             (catch Exception e (ex-data e)))]
    (with-meta body {:status status
                     :url url})))

(defn- ^:no-doc custom-fields->query-params
  "Transforms map of custom fields into query-param friendly values
  {:external_id 123} => {\"custom_fields[external_id]\" 123}"
  [fields]
  (into {} (map #(let [[k v] %]
                  [(format "custom_fields[%s]" (name k)) v]) fields)))

(defn upsert-filters
  "Transform custom-field filters to be query-param friendly"
  [filters]
  (if-let [custom-fields (get filters :custom_fields)]
    (merge (dissoc filters :custom_fields)
           (custom-fields->query-params custom-fields))
    filters))

(defn- ^:no-doc replace-resource-ids
  "Given an endpoint and a hash-map, returns a new string with all
  keys in map found in input replaced with the value of the key"
  [endpoint m]
  (reduce
    (fn [acc [k v]] (s/replace acc (str ":" (name k)) (str v)))
    endpoint m))

(def api-spec
  "The spec of the Base CRM API which is currently supported by this library."
  ;TODO - Support more resources.  Updating this hash-map with more resources will just make it work.
  ;Feel free to submit a PR!  :)
  {::leads ["/v2/leads" "/v2/leads/:id"]
   ::contacts ["/v2/contacts" "/v2/contacts/:id"]
   ::lead-sources ["/v2/lead_sources" "/v2/lead_sources/:id"]
   ::notes ["/v2/notes" "/v2/notes/:id"]})

(defn retrieve
  "Retrieve a single resource record.

   ##resource-ids

    `{:id 12345}` - Endpoints with no resource associations [(e.g. Leads)](https://developers.getbase.com/docs/rest/reference/leads)
    `{:deal_id 123 :contact_id 456}` - Endpoints with associations to other resources. [(e.g. Associated Contacts)](https://developers.getbase.com/docs/rest/reference/associated_contacts)

    [The resource-ids map to the Base API Endpoints](https://developers.getbase.com/docs/rest/articles/introduction)"
  [resource resource-ids]
  (when-let [[_ endpoint-with-ids] (get api-spec resource)]
    (base-api-call :get (replace-resource-ids endpoint-with-ids resource-ids))))

(defn retrieve-many
  "Retrieve a many resource records.  Use the params map to
   filter the records which record to retreive.

   ##resource-ids

    `{:id 12345}` - Endpoints with no resource associations [(e.g. Leads)](https://developers.getbase.com/docs/rest/reference/leads)
    `{:deal_id 123 :contact_id 456}` - Endpoints with associations to other resources. [(e.g. Associated Contacts)](https://developers.getbase.com/docs/rest/reference/associated_contacts)

    [The resource-ids map to the Base API Endpoints](https://developers.getbase.com/docs/rest/articles/introduction)"
  ([resource]
   (retrieve-many resource nil nil))
  ([resource params]
   (retrieve-many resource nil params))
  ([resource resource-ids {:keys [page per_page sort_by ids creator_id owner_id source_id] :as params}]
   (when-let [[endpoint endpoint-with-ids] (get api-spec resource)]
     (base-api-call :get (cond
                           (or (empty? resource-ids) (nil? resource-ids)) endpoint
                           (= 1 (count resource-ids)) (replace-resource-ids endpoint-with-ids resource-ids)
                           :else (replace-resource-ids endpoint-with-ids resource-ids))
                    {:query-params params}))))

(defn create
  "Create a new resource record.

   You should use the arity with resource-ids when the resource you are creating has an association to another resource. [(e.g. Associated Contacts)](https://developers.getbase.com/docs/rest/reference/associated_contacts)

   ##resource-ids

    `{:deal_id 123}` - Endpoints with associations to other resources. [(e.g. Associated Contacts)](https://developers.getbase.com/docs/rest/reference/associated_contacts)

    [The resource-ids map to the Base API Endpoints](https://developers.getbase.com/docs/rest/articles/introduction)"
  ([resource data]
   (create resource nil data))
  ([resource resource-ids data]
    ;Use this arity for resources with associations to other resources
   (when-let [[endpoint endpoint-with-ids] (get api-spec resource)]
     (base-api-call :post (if resource-ids
                            (replace-resource-ids endpoint-with-ids resource-ids)
                            endpoint)
                    {:form-params {:data data}}))))

(defn delete
  "Deletes an existing resource record.

   ##resource-ids

    `{:id 12345}` - Endpoints with no resource associations [(e.g. Leads)](https://developers.getbase.com/docs/rest/reference/leads)
    `{:deal_id 123 :contact_id 456}` - Endpoints with associations to other resources. [(e.g. Associated Contacts)](https://developers.getbase.com/docs/rest/reference/associated_contacts)

    [The resource-ids map to the Base API Endpoints](https://developers.getbase.com/docs/rest/articles/introduction)"
  [resource resource-ids]
  (when-let [[_ endpoint-with-ids] (get api-spec resource)]
    (base-api-call :delete (replace-resource-ids endpoint-with-ids resource-ids))))

(defn update
  "Updates an existing resource record.

   ##resource-ids

    `{:id 12345}` - Endpoints with no resource associations [(e.g. Leads)](https://developers.getbase.com/docs/rest/reference/leads)
    `{:deal_id 123 :contact_id 456}` - Endpoints with associations to other resources. [(e.g. Associated Contacts)](https://developers.getbase.com/docs/rest/reference/associated_contacts)

    [The resource-ids map to the Base API Endpoints](https://developers.getbase.com/docs/rest/articles/introduction)"
  [resource resource-ids data]
  (when-let [[_ endpoint-with-ids] (get api-spec resource)]
    (base-api-call :put (replace-resource-ids endpoint-with-ids resource-ids)
                   {:form-params {:data data}})))

(defn upsert
  "Create a new resource record or update an existing resource record, based on a value of a filter or a set of filters.
  At least one filter is required.

  ##Behaviour

  - If multiple records match a set of filters, the request will return an error - 409.
  - If a single record matches, then the existing record is updated
  - If no record matches the query, a new record is created.

  ##filters

  `{:email \"john.doe@fakemail.com\"}` - Filters by email
  `{:custom_fields {:external_id 123}}` - Filters by the custom field 'external_id'"
  [resource {:keys [data filters]}]
  (when-let [[endpoint _] (get api-spec resource)]
    (base-api-call :post (str endpoint "/upsert")
                   {:query-params (upsert-filters filters)
                    :form-params  {:data data}})))